File: main/java/location/Orientation.java
package location;


public enum Orientation {

    SOUTH("S", true, -1) {
        public Orientation left() {return WEST;}
        public Orientation right() {return EAST;}
    },
    WEST("W", false, -1) {
        public Orientation left() {return SOUTH;}
        public Orientation right() {return NORTH;}
    },
    NORTH("N", true, +1) {
        public Orientation left() {return WEST;}
        public Orientation right() {return EAST;}
    },
    EAST("E", false, +1) {
        public Orientation left() {return NORTH;}
        public Orientation right() {return SOUTH;}
    };


    private String string;
    public final boolean isVertical;
    public final int move;

    Orientation(String string, boolean isVertical, int move) {
        this.string = string;
        this.isVertical = isVertical;
        this.move = move;
    }

    @Override
    public String toString() {
        return this.string;
    }

    public abstract Orientation left();
    public abstract Orientation right();

}

File: main/java/com/marsrovers/location/Orientation.kt
package com.marsrovers.location


import java.util.Arrays
import java.util.Optional

enum class Orientation constructor(private val string: String, val isVertical: Boolean, val move: Int) {

    SOUTH("S", true, -1) {
        override fun left() = WEST
        override fun right() = EAST
    },
    WEST("W", false, -1) {
        override fun left() = SOUTH
        override fun right() = NORTH
    },
    NORTH("N", true, +1) {
        override fun left() = WEST
        override fun right() = EAST
    },
    EAST("E", false, +1) {
        override fun left() = NORTH
        override fun right() = SOUTH
    };

    override fun toString(): String = this.string

    abstract fun left(): Orientation
    abstract fun right(): Orientation

    companion object {

        fun fromString(orientation: String): Orientation {

            val orientationOptional = Arrays
                    .stream(values())
                    .filter { o -> o.string == orientation }
                    .findFirst()

            if (!orientationOptional.isPresent) {
                throw IllegalArgumentException("Orientation $orientation does not exists")
            }

            return orientationOptional.get()
        }
    }

}
